src/sql/src/func.rs

```diff
@@ -1437,67 +1238,307 @@ fn find_match<'a, R: std::fmt::Debug>(
     )
 }
 
-/// Coerces concrete arguments for a function according to the abstract
-/// parameters specified in the function definition.
-///
-/// You must only call this function if `ParamList::matches_argtypes` has
-/// verified that the `args` are valid for `params`.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum PolymorphicCompatClass {
+    /// Represents the older "Any"-style matching of PG polymoprhic types, which
+    /// constrains all types to be of the same type, i.e. does not attempt to
+    /// promote parameters to a best common type.
+    BaseEq,
+    /// Represent's Postgres' "anycompatible"-type polymorphic resolution.
+    ///
+    /// > Selection of the common type considers the actual types of
+    /// > anycompatible and anycompatiblenonarray inputs, the array element
+    /// > types of anycompatiblearray inputs, the range subtypes of
+    /// > anycompatiblerange inputs, and the multirange subtypes of
+    /// > anycompatiblemultirange inputs. If anycompatiblenonarray is present
+    /// > then the common type is required to be a non-array type. Once a
+    /// > common type is identified, arguments in anycompatible and
+    /// > anycompatiblenonarray positions are automatically cast to that type,
+    /// > and arguments in anycompatiblearray positions are automatically cast
+    /// > to the array type for that type.
+    ///
+    /// See https://www.postgresql.org/docs/current/extend-type-system.html#EXTEND-TYPES-POLYMORPHIC
+    BestCommonAny,
+    /// Represents polymorphic compatibility operations for Materialize LIST
+    /// types. This differs from PG's "anycompatible" type resolution, which
+    /// focuses on determining a common type, and e.g. using that as
+    /// `AnyCompatibleArray`'s element type. Instead, our list compatibility
+    /// focuses on finding a common list type, and then casting
+    /// `ListElementAnyCompatible` parameters to that list type's elements. This
+    /// approach is necessary to let us polymorphically resolve custom list
+    /// types without losing their OIDs.
+    BestCommonList,
+    /// Represents an operation similar to LIST compatibility, but for MAP. This
+    /// is distinct from `BestCommonList` in as much as the parameter types that
+    /// work with `BestCommonList` are incommensurate with the parameter types
+    /// used with `BestCommonMap`.
+    BestCommonMap,
+    /// Represents type resolution for `ScalarType::Record` types, which e.g.
+    /// ignores custom types and type modifications.
+    ///
+    /// In PG, this is handled by invocation of the function calls that take
+    /// `RecordAny` params, which we want to avoid if at all possible.
```

Comment:
What's the problem with doing the same thing in Materialize?

Reply:
Wouldn't it break row comparisons, which rely on byte-level equality?

Reply:
Oh, do those functions do something more complicated that just comparing element by element?

Reply:
```
sean=# SELECT ROW(1, 2) = ROW(1, NULL);
 ?column? 
----------
 
(1 row)
```

I guess that could get lowered into dataflow; was a little bit more than I had considered committing to given that we _can_ just type check in planning. Or mostly, I think--is there something in particular you see as an issue?

Reply:
Ok--I need to spend some time digging into the equality semantics for these types to figure out if this could get pushed down elsehwere or not:

```
sean=#  CREATE TYPE c AS (i int);
sean=# SELECT ROW(NULL)::c = ROW(NULL::int);
 ?column?
----------
 t
(1 row)

sean=# SELECT ROW(NULL) = ROW(NULL::int);
 ?column?
----------

```